appcelerator/titanium_mobile
# [TIMOB-24963] iOS/Android: Added new ScrollView.scrollToTop() method

**Note:**
This is a modified version of @m1ga's PR #9221. (Thanks Michael!)

**JIRA:**
_Main Ticket_
https://jira.appcelerator.org/browse/TIMOB-24963

_Also Fixes_
https://jira.appcelerator.org/browse/TIMOB-25514
https://jira.appcelerator.org/browse/TIMOB-25515

**Summary:**
- Added new ScrollView.scrollToTop() method.
- [TIMOB-25514] Fixes Android bug where vertical ScrollView.scrollTo() goes to wrong position when animated. (Introduced in Titanium 6.2.2.)
- [TIMOB-25515] Fixes Android bug where vertical ScrollView.scrollTo() ignores "ti.ui.defaultunit" when animated.
- Animated vertical scrolling bugs mentioned above were solved by disabling the smooth scrolling animation. It now immediately jumps to the given position. This appears to be a Google bug in their Java `NestedScrollView` class.

**Test:**

Run the below test on both Android and iOS...

1. Build and run the below code on an Android or iOS device.
2. An alert dialog will be displayed. Tap the "Vertical" scroll view button.
2. Tap the scroll-to "End" button.
3. Verify that the scroll view scrolls to the bottom.
4. Tap the scroll-to "Middle" button.
5. Verify that the top of the middle view is now displayed at the top of the scroll view.
6. Tap the scroll-to "Beginning" button.
7. Verify that the scroll view scrolls to the top.
8. Restart the app.
9. An alert dialog will be displayed. Tap the "Horizontal" scroll view button.
10. Tap the scroll-to "End" button.
11. Verify that the scroll view scrolls all the way to the right side.
12. Tap the scroll-to "Middle" button.
13. Verify that the left side of the middle view is now displayed at the left edge of the scroll view.
14. Tap the scroll-to "Beginning" button.
15. Verify that the scroll view scrolls all the way to the left side.

```javascript
function showScrollView(isVertical) {
	var window = Ti.UI.createWindow();
	var scrollView = Ti.UI.createScrollView(
	{
		layout: isVertical ? "vertical" : "horizontal",
		scrollType: isVertical ? "vertical" : "horizontal",
		showHorizontalScrollIndicator: !isVertical,
		shorVerticalScrollIndicator: isVertical,
		width: Ti.UI.FILL,
		height: Ti.UI.FILL,
	});
	var firstView = Ti.UI.createLabel(
	{
		text: "First View",
		textAlign: Ti.UI.TEXT_ALIGNMENT_CENTER,
		color: "white",
		backgroundColor: "blue",
		borderColor: "white",
		borderWidth: "1dp",
		width: isVertical ? "100%" : "75%",
		height: isVertical ? "75%" : "100%",
	});
	scrollView.add(firstView);
	var middleView = Ti.UI.createLabel(
	{
		text: "Middle View",
		textAlign: Ti.UI.TEXT_ALIGNMENT_CENTER,
		color: "white",
		backgroundColor: "#008800",
		borderColor: "white",
		borderWidth: "1dp",
		width: isVertical ? "100%" : "75%",
		height: isVertical ? "75%" : "100%",
	});
	scrollView.add(middleView);
	var lastView = Ti.UI.createLabel(
	{
		text: "Last View",
		textAlign: Ti.UI.TEXT_ALIGNMENT_CENTER,
		color: "white",
		backgroundColor: "purple",
		borderColor: "white",
		borderWidth: "1dp",
		width: isVertical ? "100%" : "75%",
		height: isVertical ? "75%" : "100%",
	});
	scrollView.add(lastView);
	window.add(scrollView);
	var floatingView = Ti.UI.createView(
	{
		layout: "vertical",
		backgroundColor: "white",
		borderColor: "black",
		borderWidth: "1dp",
		bottom: "10dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	floatingView.add(Ti.UI.createLabel(
	{
		text: "Scroll to:",
		color: "black",
		top: "5dp",
		bottom: "5dp",
		left: "10dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	}));
	var buttonContainer = Ti.UI.createView(
	{
		layout: "horizontal",
		bottom: "5dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	var scrollUpButton = Ti.UI.createButton(
	{
		title: "Beginning",
		left: "5dp",
		right: "5dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	scrollUpButton.addEventListener("click", function(e) {
		scrollView.scrollToTop();
	});
	buttonContainer.add(scrollUpButton);
	var scrollMiddleButton = Ti.UI.createButton(
	{
		title: "Middle",
		left: "5dp",
		right: "5dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	scrollMiddleButton.addEventListener("click", function(e) {
		var x = isVertical ? 0 : firstView.size.width;
		var y = isVertical ? firstView.size.height : 0;
		scrollView.scrollTo(x, y, { animated: true });
	});
	buttonContainer.add(scrollMiddleButton);
	var scrollDownButton = Ti.UI.createButton(
	{
		title: "End",
		left: "5dp",
		right: "5dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	scrollDownButton.addEventListener("click", function(e) {
		scrollView.scrollToBottom();
	});
	buttonContainer.add(scrollDownButton);
	floatingView.add(buttonContainer);
	window.add(floatingView);
	window.add(buttonContainer);
	window.open();
}

var dialog = Ti.UI.createAlertDialog(
{
	message: "Select ScrollView Type",
	buttonNames: ["Vertical", "Horizontal"],
});
dialog.addEventListener("click", function(e) {
	if (e.index == 0) {
		showScrollView(true);
	} else if (e.index == 1) {
		showScrollView(false);
	}
});
if ((Ti.Platform.osname === "iphone") || (Ti.Platform.osname === "ipad")) {
	var rootWindow = Ti.UI.createWindow();
	rootWindow.addEventListener("open", function(e) {
		dialog.show();
	});
	rootWindow.open();
} else {
	dialog.show();
}
```
